sql/select.go

```diff
@@ -85,3 +99,523 @@ func (p *planner) expandSubqueries(stmt parser.Statement) error {
+			if s.index == index.index {
+				s.startKey = index.makeStartKey()
+				s.endKey = index.makeEndKey()
+				return s, nil
+			}
+		}
+	}
+
+	// After sorting, indexes[0] contains the best index. Copy its info into the
+	// scanNode.
+	s.index = indexes[0].index
+	s.isSecondaryIndex = (s.index != &s.desc.PrimaryIndex)
+	s.startKey = indexes[0].makeStartKey()
+	s.endKey = indexes[0].makeEndKey()
+	return s, nil
+}
+
+// qvalueInfo contains one end of a value range.
+type qvalueInfo struct {
+	datum parser.Datum
+	op    parser.ComparisonOp
+}
+
+func (q *qvalueInfo) assertStartOp() {
+	switch q.op {
+	case parser.EQ, parser.GT:
+	default:
+		panic(fmt.Sprintf("unexpected start op: %s", q.op))
+	}
+}
+
+func (q *qvalueInfo) assertEndOp() {
+	switch q.op {
+	case parser.EQ, parser.LT:
+	default:
+		panic(fmt.Sprintf("unexpected end op: %s", q.op))
+	}
+}
+
+func (q *qvalueInfo) compare(datum parser.Datum) int {
+	// TODO(pmattis): We should really be comparing the datums directly, but this
+	// is convenient for now.
+	oldKey, err := encodeTableKey(nil, q.datum)
+	if err != nil {
+		panic(err)
+	}
+	newKey, err := encodeTableKey(nil, datum)
+	if err != nil {
+		panic(err)
+	}
+	return bytes.Compare(newKey, oldKey)
+}
+
+// intersectStart intersects the start value information.
+func (q *qvalueInfo) intersectStart(n qvalueInfo) {
+	n.assertStartOp()
+
+	if q.datum == nil || q.compare(n.datum) > 0 {
+		// The new datum is larger than the existing one.
+		q.datum = n.datum
+		q.op = n.op
+	}
+}
+
+// intersectEnd intersects the start value information.
+func (q *qvalueInfo) intersectEnd(n qvalueInfo) {
+	n.assertEndOp()
+
+	if q.datum == nil || q.compare(n.datum) < 0 {
+		// The new datum is smaller than the existing one.
+		q.datum = n.datum
+		q.op = n.op
+	}
+}
+
+// unionStart unions the start value information.
+func (q *qvalueInfo) unionStart(n qvalueInfo) {
+	n.assertStartOp()
+
+	if q.datum == nil {
+		q.datum = n.datum
+		q.op = n.op
+		return
+	}
+
+	switch q.compare(n.datum) {
+	case -1:
+		// The new datum is smaller than the existing one.
+		q.datum = n.datum
+		q.op = n.op
+	case 0:
+		if n.op == parser.EQ {
+			q.op = parser.EQ
+		}
+	}
+}
+
+// unionEnd unions the start value information.
+func (q *qvalueInfo) unionEnd(n qvalueInfo) {
+	n.assertEndOp()
+
+	if q.datum == nil {
+		q.datum = n.datum
+		q.op = n.op
+		return
+	}
+
+	switch q.compare(n.datum) {
+	case +1:
+		// The new datum is larger than the existing one.
+		q.datum = n.datum
+		q.op = n.op
+	case 0:
+		if n.op == parser.EQ {
+			q.op = parser.EQ
+		}
+	}
+}
+
+// qvalueRange represents the range of values a qvalue may have.
+type qvalueRange struct {
+	start qvalueInfo
+	end   qvalueInfo
+}
+
+type qvalueRangeMap map[ColumnID]*qvalueRange
+
+// analyzeExpr walks over the specified expression and populates the range map
+// with the value ranges for each qvalue.
+func (m qvalueRangeMap) analyzeExpr(expr parser.Expr) {
+	switch t := expr.(type) {
+	case *parser.NotExpr:
```

Comment:
I don't think so. Not implementing the analysis for an expression simply means we'll fall back to a table scan.